Adding taxi-ride implementation in Akka

blublinsky opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/lightbend/cloudflow/blob/master/CONTRIBUTING.md
  2. Ensure you have added or run the appropriate tests for your PR.
  3. If the PR is unfinished, open it as a 'Draft'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
This is just a sample showing how to implement taxi ride functionality in Akka Streams
-->


### Why are the changes needed?
<!--
This example shows implementation of the same functionality as in Flink, in Akka. This allows to make a better decision on runtime to use
-->


### Does this PR introduce any user-facing change?
<!--
'No'.
-->


### How was this patch tested?
<!--
It was tested through run local
-->

RayRoestenburg commented:
Is this a draft PR @blublinsky ?

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+## Akka implementation of taxi ride application
+
+This example demonstrates how `taxi-ride` application initially implemented in flink can be implemented using Akka.
```
```suggestion
This example demonstrates how the `taxi-ride` application, initially implemented in Flink, can be implemented using Akka.
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+This example demonstrates how `taxi-ride` application initially implemented in flink can be implemented using Akka.
+
+The implementation combines two streams `TaxiRide` and `TaxiFare` and for each rideID calculates
+a fare for a given ride 
```
```suggestion
a fare for a given ride. 
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+a fare for a given ride 
+
+In the case of Flink we used `RichCoFlatMapFunction` to combine the two streams, here
+we are using `RideShare` actor for the same purpose. If we compare implementation of the actor 
```
```suggestion
we are using a `RideShare` actor for the same purpose. If we compare implementation of the actor 
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+we are using `RideShare` actor for the same purpose. If we compare implementation of the actor 
+with the implementation of Flink's `RichCoFlatMapFunction`, we will see that they are very similar.
+The missing part here is persistence. While Flink implementation provides automatic snapshotting, Akka
+requires implementation of an explicit persistence support, which is not currently supported
```
Why is this not supported? Do you mean we do not install a persistent journal? It would be great if that can be done in this example. (is there a simple option that would work on a distributed file system, using a pvc?) Otherwise it's not a great comparison.

RayRoestenburg replied:
Please add persistence for jdbc, with a postgres helm chart for instance.

blublinsky replied:
Done

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+requires implementation of an explicit persistence support, which is not currently supported
+by Cloudflow (but can be easily added manually).
+
+Main flow of Flink implementation is very elegant and rasy to read   
```
```suggestion
Main flow of Flink implementation is very elegant and easy to read:
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
`````diff
@@ -0,0 +1,95 @@
+````
+Note that here there is no explicit `keyBy` (we will discuss it later). 
+Unlike Flink implementation, where streams are connected before invoking, in the case of
+Akka, we process each stream independantly and then merge results afterwards.
`````
```suggestion
Akka, we process each stream independently and then merge results afterwards.
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+So the implementation provides the same functionality (except for automatic persistence, 
+which can be easily added), but with less user friendly syntax. Flink's implementation hides
+all of the sharding details by usage of `keyBy`.
+On another hand, FLink uses static topology for applications, which means that scaling requires
```
```suggestion
On the other hand, Flink uses a static topology for applications, which means that scaling requires
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+which can be easily added), but with less user friendly syntax. Flink's implementation hides
+all of the sharding details by usage of `keyBy`.
+On another hand, FLink uses static topology for applications, which means that scaling requires
+application restarts, while in the case of Akka, adding additional instance can be done without 
```
What do you mean here, that existing, running actor systems as part of the cluster don't have to restart?

blublinsky replied:
exactly. You can add an additional instance without any restarts of the running ones

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+application restarts, while in the case of Akka, adding additional instance can be done without 
+any restarts.
+
+The interesting feature of this example, is that the keys `rideID` do not belong to a fix set,
```
```suggestion
The interesting feature of this example, is that the keys `rideID` do not belong to a fixed set,
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+
+The interesting feature of this example, is that the keys `rideID` do not belong to a fix set,
+but rather change all the time. As a result Akka implementation creates and deletes actors
+all the time. FLink alleviates this problem by using keys sharding - using a single implementation
```
```suggestion
all the time. Flink alleviates this problem by using keys sharding - using a single implementation
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+is leveraging [Kafka aware sharding](https://doc.akka.io/docs/alpakka-kafka/current/cluster-sharding.html),
+which means that each instance processes a set of partitions from the topic. In the case of scaling,
+the partition as a whole can be moved to a different instance, but can not be split.
+This means that we can use actors based on partitions to avoid constant acters creation and deletion.
```
```suggestion
This means that we can use actors based on partitions to avoid constant actor creation and deletion.
```

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+we can use a single topic for both messages.
+* Default Kafka [partitioner](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java)
+uses a 32-bit murmur2 hash to compute the partition id based on the key (bytes and the amount of partition).
+If we assume that the defalt partitioner is used, we can implement key calculation as following:
```
```suggestion
If we assume that the default partitioner is used, we can implement key calculation as following:
```

RayRoestenburg replied:
Do you need to assume? Or is this just fact?

RayRoestenburg commented on examples/taxi-ride-akka/akkastreams/src/main/scala/taxiride/akka/processor/actors/RidesBehaviour.scala:
```diff
@@ -0,0 +1,51 @@
+case class ProcessRide(reply: ActorRef[Option[TaxiRideFare]], ride : TaxiRide) extends RidesProcessingActor
+
+// Actor
+object RideShare{
```
Not formatted, please add scalafmt?

blublinsky replied:
done

RayRoestenburg commented on examples/taxi-ride-akka/akkastreams/src/main/scala/taxiride/akka/processor/actors/RidesBehaviour.scala:
```diff
@@ -0,0 +1,51 @@
+
+  private def processMessage(rideid: String): Behavior[RidesProcessingActor] =
+     Behaviors.receive { (context, msg) => {
+//       context.log.info(s"a new message $msg for ride $rideid")
```
Remove commented lines please. (nitpick)

blublinsky replied:
Done

RayRoestenburg commented on examples/taxi-ride-akka/taxi-ride-pipeline/src/main/blueprint/blueprint.conf.txt:
```diff
@@ -0,0 +1,25 @@
+blueprint {
```
remove this file?

blublinsky replied:
I have 2 blueprints for two different use cases

RayRoestenburg commented on examples/taxi-ride-akka/README.md:
```diff
@@ -0,0 +1,95 @@
+the partition as a whole can be moved to a different instance, but can not be split.
+This means that we can use actors based on partitions to avoid constant acters creation and deletion.
+
+To achieve this we need to make several changes to our implementation.
```
It's probably better to just show one solution (the single message) and describe the reasoning for it when using Akka (colocating shard and consumer).

blublinsky replied:
i like options, sorry

blublinsky replied:
There are different use cases and both approaches have their sweet spots

RayRoestenburg commented on examples/taxi-ride-akka/akkastreams/src/main/scala/taxiride/akka/processor/actors/singlemessage/RidesBehaviour.scala:
```diff
@@ -0,0 +1,47 @@
+import taxiride.datamodel.TaxiRideOrFare.MessageType
+import taxiride.datamodel._
+
+import scala.collection.mutable.Map
```
You can't just use a mutable map in an object.

blublinsky replied:
Actually you can. All of the actors are **_local_**. But doing it cleaner in rework

RayRoestenburg replied:
(it's not about it being local, I'm referring to concurrency issues, but good to know it will be reworked)

RayRoestenburg commented on examples/taxi-ride-akka/akkastreams/src/main/scala/taxiride/akka/processor/actors/singlemessage/RidesBehaviour.scala:
```diff
@@ -0,0 +1,47 @@
+object RideShare{
+
+  private val rideState: Map[Long, TaxiRide] = Map()
+  private val fareState: Map[Long, TaxiFare] = Map()
```
This needs more work.

RayRoestenburg commented on examples/taxi-ride-akka/akkastreams/src/main/scala/taxiride/akka/processor/actors/RidesBehaviour.scala:
```diff
@@ -0,0 +1,51 @@
+object RideShare{
+
+  private val rideState: Map[Long, TaxiRide] = Map()
+  private val fareState: Map[Long, TaxiFare] = Map()
```
This needs more work, see further down.

blublinsky replied:
There are options here. Let me finish the first path through of Flink

RayRoestenburg reviewed (changes requested):
This example needs persistent actors (using jdbc journal preferably)

blublinsky commented:
Ok, Significantly reworked code, added persistence and reworked readme.
Once you are done with review, I would like to move some of the things to cloudflow core, namely:
1. Add required jars
2. Add CborSerializable
3. Add extended application.conf
